### DISCUSSION TITLE
How to type a reusable resolver in Typescript

### DISCUSSION THREAD

**Discussion by javipuche** [mswjs/msw#1719] [Q&A] [answered]
Hello, I have some questions about typing when using a reusable resolver. Is it something like this?:

```ts
import type { ResponseResolver } from 'msw'

export const mockUser: ResponseResolver = (req, res, ctx) => {
  return res(
    ctx.json({
      firstName: 'John',
      age: 38,
    }),
  )
}
```

This show an error in `ctx.json`. The property `json` doesn't exists in `DefaultContext`. 

Or maybe, Is it something like this? (I don't know what to use for "res" in this case):

```ts
import type { RestContext, RestRequest } from 'msw'

export const mockUser = (req: RestRequest, res, ctx: RestContext) => {
  return res(
    ctx.json({
      firstName: 'John',
      age: 38,
    }),
  )
}
```

Thank you for clearing up my doubts

**Comment by kettanaito** [comment, marked as answer]
Hi, @javipuche. Thanks for raising this. 

## Getting to know `ResponseResolver` better

Take a look at the `ResponseResolver` type:

https://github.com/mswjs/msw/blob/ad27719ac375301ebd2b736cddb86cb5061a4f26/src/handlers/RequestHandler.ts#L73-L81

It accepts three generics:

- `RequestType`, this is the type for the `req` provided to the response resolver function.
- `ContextType`, the type of the `ctx` object.
- `BodyType`, the _response_ body type to narrow down what responses `res` function should compose.

The way MSW works is that there are common `ctx` utilities, and then utilities specific to particular request types: REST (like `ctx.json()` and `ctx.set()`) and GraphQL (like `ctx.data()`). By default, `ResponseResolver` doesn't know what kind of requests you mean to intercept with it, so you have to explicitly provide that on the type level. 

## How to fix this?

MSW ships with pre-defined request and context types you can reuse directly. For example, if you are annotating your custom resolver that's meant to handle REST API requests, use this type:

```js
import type { ResponseResolver, RestRequest, RestContext } from 'msw'

interface User {
  firstName: string
  age: number
}

export const mockUser: ResponseResolver<RestRequest, RestContext, User> = (req, res, ctx) => {
  return res(
    ctx.json({
      firstName: 'John',
      age: 38,
    }),
  )
}
```

## Learn more

Looking at MSW's source code is a great place to see how to use these types, for example:

https://github.com/mswjs/msw/blob/ad27719ac375301ebd2b736cddb86cb5061a4f26/src/rest.ts#L19-L28

https://github.com/mswjs/msw/blob/ad27719ac375301ebd2b736cddb86cb5061a4f26/src/graphql.ts#L34-L37

You opening this discussion also helped me realize that the [Type-safe API Mocking With Mock Service Worker](https://dev.to/kettanaito/type-safe-api-mocking-with-mock-service-worker-and-typescript-21bf) article I wrote a while ago has an incorrect example of annotating custom response resolver. It's fixed now! Feel free to refer to that article for more examples on how to use MSW with TypeScript. Hope this helps.

**Reply by javipuche** [reply]
Thank you very much for the assistance and for the explanation. I will take a look at the article ;)
